### PR TITLE
8.0 Update for Choose a web UI

### LINKED ISSUE #30497: Revise guidance for .NET 8 to recommend Blazor for web UI
With .NET 8 our guidance for choosing a ASP.NET Core web UI technology gets much simpler: You should just use Blazor. Blazor in .NET 8 can now do static server rendering (like MVC & Razor Pages), interactive server rendering (with Blazor Server), and interactive client-side rendering (with Blazor WebAssembly). So, you can really do it all from within one web UI framework. Blazor also has a rich component model that enables reusability and consistency, and it has new advanced rendering capabilities for progressively enhancing the user experience.

We should revise this topic to lead with Blazor as the recommend web UI approach. We should follow that with mentions of MVC & Razor Pages as mature and supported alternatives. We can mention that Blazor can be integrated with existing MVC & Razor Pages apps if desired. And we should still mention that pairing ASP.NET Core with frontend JavaScript frameworks is also supported. But Blazor should be the hero here.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 21671480-fba7-3692-1a20-e71843864b60
* Version Independent ID: 68d5aab0-63ce-de7a-0c56-341c2eef623d
* Content: [Choose an ASP.NET Core UI](https://learn.microsoft.com/en-us/aspnet/core/tutorials/choose-web-ui?view=aspnetcore-7.0)
* Content Source: [aspnetcore/tutorials/choose-web-ui.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/tutorials/choose-web-ui.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-tutorials**
* GitHub Login: @wadepickett
* Microsoft Alias: **wpickett**

---
[Associated WorkItem - 164336](https://dev.azure.com/msft-skilling/Content/_workitems/edit/164336)

### PR DESCRIPTION
**Pull request by wadepickett**
[Internal review: Choose an ASP.NET Core web UI](https://review.learn.microsoft.com/en-us/aspnet/core/tutorials/choose-web-ui?view=aspnetcore-8.0&branch=pr-en-us-30569)

Fixes #30497

- Revised the Choose a web UI topic to lead with Blazor as the recommend web UI approach.
- Added mentions that MVC & Razor Pages as mature and supported alternatives to Blazor.
- Removed the server-render vs client render content since you no longer must choose. With Blazor in .NET 8 you get both from the same app (and more).
- Verified we still mention that Blazor can be integrated with existing MVC & Razor Pages apps if desired. We do, see last section _Choose a hybrid solution: ASP.NET Core MVC or Razor Pages plus Blazor_,

I am reusing some key Blazor introductions already approved and published for .NET 8 from guardrex, Daniel Roth and others for the 8.0 update of the Blazor description here with a few changes to make it work here.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/tutorials/choose-web-ui.md](https://github.com/dotnet/AspNetCore.Docs/blob/828a060416d31311e245906bf70a42a42a6cbd8c/aspnetcore/tutorials/choose-web-ui.md) | [Choose an ASP.NET Core web UI](https://review.learn.microsoft.com/en-us/aspnet/core/tutorials/choose-web-ui?branch=pr-en-us-30569) |

### CODE REVIEW

**Comment by wadepickett** [comment]
@guardrex, I'm using much of your text and Dan's from 8.0 Blazor introductions here to lead with a Blazor intro and benefits list that fits the scope of this topic with a few changes.  I'm hoping you can take a look and let me know if I'm off here with anything.

The newly titled section: _Blazor: the recommended ASP.NET full-stack web UI_ might be a bit too bold.  Since we have a recommendation, considering the topic, I think it should be really fast to find and clear, but I could move it out of the section title if that seems too much.

File: aspnetcore/tutorials/choose-web-ui.md [hunk 1]
```diff
@@ -3,83 +3,55 @@ title: Choose an ASP.NET Core UI
...
-There are three general approaches to building modern web UI with ASP.NET Core:
+Blazor provides an all-in-one web app solution and is recommended as a first choice for most ASP.NET web UI scenarios, starting with .NET 8.
 
-* Apps that render UI from the server.
-* Apps that render UI on the client in the browser.
-* Hybrid apps that take advantage of both server and client UI rendering approaches. For example, most of the web UI is rendered on the server, and client rendered components are added as needed.
+Blazor is a full-stack web UI framework for developing apps that renders content at a reusable component or page level with:
 
-There are benefits and drawbacks to consider when rendering UI on the server or on the client.
+* Switchable rendering modes that can be chosen at runtime for server side or client side interactivity.
```
Comment (guardrex) [inline aspnetcore/tutorials/choose-web-ui.md:19]:
```suggestion
* Switchable rendering modes that can be chosen at runtime for server-side or client-side interactivity.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 2]
```diff
@@ -3,83 +3,55 @@ title: Choose an ASP.NET Core UI
...
+* Blazor WebAssembly bootstraps the .NET runtime and configures the runtime to load the assemblies for the app. The Blazor WebAssembly runtime uses JavaScript interop to handle DOM manipulation and browser API calls.
 
-## Choose a server rendered ASP.NET Core UI solution
+For a complete overview Blazor, its architecture and benefits, see <xref:blazor/index> and <xref:blazor/hosting-models>. To get started with your first Blazor app, see [Build your first Blazor app](https://dotnet.microsoft.com/learn/aspnet/blazor-tutorial/intro).
 
-The following section explains the ASP.NET Core web UI server rendered models available and provides links to get started. ASP.NET Core Razor Pages and ASP.NET Core MVC are server-based frameworks for building web apps with .NET.
+## ASP.NET Core Razor Pages
 
-### ASP.NET Core Razor Pages
+Razor Pages are a mature and supported alternative to Blazor. Razor Pages is a page-based model. UI and business logic concerns are kept separate, but within the page.
```
Comment (guardrex) [inline aspnetcore/tutorials/choose-web-ui.md:50]:
```suggestion
Razor Pages are a mature and supported alternative to Blazor. Razor Pages adopts a page-based model. UI and business logic concerns are kept separate, but within the page.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 3]
```diff
@@ -3,83 +3,55 @@ title: Choose an ASP.NET Core UI
...
 
-Razor Pages is a page-based model. UI and business logic concerns are kept separate, but within the page. Razor Pages is the recommended way to create new page-based or form-based apps for developers new to ASP.NET Core. Razor Pages provides an easier starting point than ASP.NET Core MVC.
+Razor pages UI are dynamically rendered on the server to generate the page's HTML and CSS on the server in response to a browser request. The page arrives at the client ready to display. Support for Razor Pages is built on ASP.NET Core MVC.
```
Comment (guardrex) [inline aspnetcore/tutorials/choose-web-ui.md:52]:
```suggestion
Razor pages UI are dynamically rendered on the server to generate the page's HTML and CSS in response to a browser request. The page arrives at the client ready to display. Support for Razor Pages is built on ASP.NET Core MVC.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 4]
```diff
@@ -89,60 +61,19 @@ Razor Pages benefits, in addition to the server rendering benefits:
...
 
-### ASP.NET Core MVC
+## ASP.NET Core MVC
 
-ASP.NET MVC renders UI on the server and uses a Model-View-Controller (MVC) architectural pattern. The MVC pattern separates an app into three main groups of components: Models, Views, and Controllers. User requests are routed to a controller. The controller is responsible for working with the model to perform user actions or retrieve results of queries. The controller chooses the view to display to the user, and provides it with any model data it requires. Support for Razor Pages is built on ASP.NET Core MVC.
+ASP.NET Core MVC is a mature and supported alternative to Blazor. ASP.NET Core MVC renders UI on the server and uses a Model-View-Controller (MVC) architectural pattern. The MVC pattern separates an app into three main groups of components: Models, Views, and Controllers. User requests are routed to a controller. The controller is responsible for working with the model to perform user actions or retrieve results of queries. The controller chooses the view to display to the user, and provides it with any model data it requires. Support for Razor Pages is built on ASP.NET Core MVC.
```
Comment (guardrex) [inline aspnetcore/tutorials/choose-web-ui.md:66]: I think the parts of the MVC app aren't proper nouns.

```suggestion
ASP.NET Core MVC is a mature and supported alternative to Blazor. ASP.NET Core MVC renders UI on the server and uses a Model-View-Controller (MVC) architectural pattern. The MVC pattern separates an app into three main groups of components: models, views, and controllers. User requests are routed to a controller. The controller is responsible for working with the model to perform user actions or retrieve results of queries. The controller chooses the view to display to the user and provides it with any model data it requires.
```

**Review by guardrex — APPROVED** [review]
Looks great! I just had a few NITs. Dan is really expert at marketing language, so he'll be of more help here.

File: aspnetcore/tutorials/choose-web-ui.md [hunk 5]
```diff
@@ -3,83 +3,55 @@ title: Choose an ASP.NET Core UI
...
-ms.date: 09/12/2023
+ms.date: 10/04/2023
 uid: tutorials/choose-web-ui
 ---
 # Choose an ASP.NET Core web UI
 
 ASP.NET Core is a complete UI framework. Choose which functionalities to combine that fit the app's web UI needs.
 
-## Benefits vs. costs of server and client rendered UI
+## Blazor: the recommended ASP.NET full-stack web UI
```
[the remark this answers is not in the document]
Comment (danroth27) [reply aspnetcore/tutorials/choose-web-ui.md:13]: How about this?:

```suggestion
## Full-stack web UI with Blazor
```

**Comment by wadepickett** [comment]
@guardrex, much better!  Thanks!

**Comment by wadepickett** [comment]
@danroth27, this seems like it is in a solid state, but we will wait for your review before publishing.

File: aspnetcore/tutorials/choose-web-ui.md [hunk 6]
```diff
@@ -3,83 +3,56 @@ title: Choose an ASP.NET Core UI
...
 ---
 # Choose an ASP.NET Core web UI
 
 ASP.NET Core is a complete UI framework. Choose which functionalities to combine that fit the app's web UI needs.
 
-## Benefits vs. costs of server and client rendered UI
+## ASP.NET Core Blazor
 
-There are three general approaches to building modern web UI with ASP.NET Core:
+Blazor is a full-stack web app framework, recommended for most ASP.NET web UI scenarios.
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:15]:
```suggestion
Blazor is a full-stack web UI framework and is recommended for most web UI scenarios.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 7]
```diff
@@ -3,83 +3,56 @@ title: Choose an ASP.NET Core UI
...
 
-* Apps that render UI from the server.
-* Apps that render UI on the client in the browser.
-* Hybrid apps that take advantage of both server and client UI rendering approaches. For example, most of the web UI is rendered on the server, and client rendered components are added as needed.
+Blazor renders content with:
 
-There are benefits and drawbacks to consider when rendering UI on the server or on the client.
+* Reusable components included in pages or whole-page, routable components.
+* Switchable rendering modes that can be chosen at runtime for server-side or client-side interactivity.
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:20]: This bullet seems a bit redundant with the following ones. Maybe just remove it?

```suggestion
* Switchable rendering modes that can be chosen at runtime for server-side or client-side interactivity.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 8]
```diff
@@ -3,83 +3,56 @@ title: Choose an ASP.NET Core UI
...
+* Static server rendering to generate static HTML.
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:21]: I'd like to also mention the progressive enhancements to static server rendering that we support:

```suggestion
* Static server rendering to generate static HTML.
* Progressive enhancements to static server rendering that enhance navigation and form handling and enable streaming rendering.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 9]
```diff
@@ -3,83 +3,56 @@ title: Choose an ASP.NET Core UI
...
 
-The following section explains the ASP.NET Core web UI server rendered models available and provides links to get started. ASP.NET Core Razor Pages and ASP.NET Core MVC are server-based frameworks for building web apps with .NET.
+## ASP.NET Core Razor Pages
 
-### ASP.NET Core Razor Pages
+Razor Pages are a mature and supported alternative to Blazor. Razor Pages adopts a page-based model. UI and business logic concerns are kept separate, but within the page.
 
-Razor Pages is a page-based model. UI and business logic concerns are kept separate, but within the page. Razor Pages is the recommended way to create new page-based or form-based apps for developers new to ASP.NET Core. Razor Pages provides an easier starting point than ASP.NET Core MVC.
+Razor pages UI are dynamically rendered on the server to generate the page's HTML and CSS in response to a browser request. The page arrives at the client ready to display. Support for Razor Pages is built on ASP.NET Core MVC.
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:53]: Even though Razor Pages is plural, I think we use it as a proper noun to refer to the singular Razor Pages framework.

I'd also prefer to not refer to Razor Pages as a Blazor alternative. Let's just talk about the strengths and weaknesses of each option.

```suggestion
Razor Pages is a page-based model for building server rendered web UI. Razor pages UI are dynamically rendered on the server to generate the page's HTML and CSS in response to a browser request. The page arrives at the client ready to display. Support for Razor Pages is built on ASP.NET Core MVC.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 10]
```diff
@@ -3,83 +3,56 @@ title: Choose an ASP.NET Core UI
...
-ms.date: 09/12/2023
+ms.date: 10/04/2023
 uid: tutorials/choose-web-ui
 ---
 # Choose an ASP.NET Core web UI
 
 ASP.NET Core is a complete UI framework. Choose which functionalities to combine that fit the app's web UI needs.
 
-## Benefits vs. costs of server and client rendered UI
+## ASP.NET Core Blazor
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:13]: The updated Blazor section seems to be focused on giving an overview of what Blazor is all about, which we already have covered in the [updated Blazor Overview for .NET 8](https://learn.microsoft.com/aspnet/core/blazor/?view=aspnetcore-8.0). This is a solid start. The new content has the high-level structure we want. I like how the Razor Pages, MVC, and JavaScript sections are structured: they briefly describe what the framework is and then list the benefits for choosing that option. Can we make the Blazor section more like the other sections?

File: aspnetcore/tutorials/choose-web-ui.md [hunk 11]
```diff
@@ -3,83 +3,56 @@ title: Choose an ASP.NET Core UI
...
+Running .NET code inside web browsers is made possible by [WebAssembly](https://webassembly.org) (abbreviated `wasm`). WebAssembly is a compact bytecode format optimized for fast download and maximum execution speed. WebAssembly is an open web standard and supported in web browsers without plugins. Blazor WebAssembly works in all modern web browsers, including mobile browsers.
 
-Drawbacks:
+When a Blazor WebAssembly is built and run:
 
-* Code for the logic has to be downloaded and executed on the client, adding to the initial load time.
-* Client requirements may exclude users who have low-end devices, older browser versions, or low-bandwidth connections.
+* C# code files and Razor files are compiled into .NET assemblies.
+* The assemblies and the [.NET runtime](/dotnet/framework/get-started/overview) are downloaded to the browser.
+* Blazor WebAssembly bootstraps the .NET runtime and configures the runtime to load the assemblies for the app. The Blazor WebAssembly runtime uses JavaScript interop to handle DOM manipulation and browser API calls.
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:45]: This is a lot of content on how Blazor can run on WebAssembly. I don't think we really need any of this in this doc.

File: aspnetcore/tutorials/choose-web-ui.md (hunk 9 above)
Comment (wadepickett) [reply aspnetcore/tutorials/choose-web-ui.md:54]: Sounds good.  I will re-edit and commit.  GitHub doesn't like that we are adding to a deleted line.

File: aspnetcore/tutorials/choose-web-ui.md (hunk 7 above)
Comment (wadepickett) [reply aspnetcore/tutorials/choose-web-ui.md:20]: I will remove it.  GitHub is giving an error on this line so I will re-commit outside this one.

File: aspnetcore/tutorials/choose-web-ui.md (hunk 10 above)
Comment (danroth27) [reply aspnetcore/tutorials/choose-web-ui.md:13]: Benefits of using Blazor:
- Reusable component model
- Efficient diff-based component rendering
- Flexibly render components from the client or server
- Build rich interactive web UI components in C#
- Render components statically from the server
- Progressively enhance server-rendered components for smoother navigation and form handling and to enable streaming rendering
- Share code for common logic on the client and server
- Interop with JavaScript
- Integrate components with existing MVC, Razor Pages, or JavaScript based apps

File: aspnetcore/tutorials/choose-web-ui.md [hunk 12]
```diff
@@ -3,83 +3,57 @@ title: Choose an ASP.NET Core UI
...
-## Benefits vs. costs of server and client rendered UI
+## ASP.NET Core Blazor
 
-There are three general approaches to building modern web UI with ASP.NET Core:
+Blazor is a full-stack web UI framework and is recommended for most web UI scenarios.
 
-* Apps that render UI from the server.
-* Apps that render UI on the client in the browser.
-* Hybrid apps that take advantage of both server and client UI rendering approaches. For example, most of the web UI is rendered on the server, and client rendered components are added as needed.
+Blazor renders content with:
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:17]: I've reviewed this list about how Blazor renders content, but I'm not really sure that we need it here. I think what we really want is a list of Blazor benefits like we have in the other sections.

File: aspnetcore/tutorials/choose-web-ui.md [hunk 13]
```diff
@@ -89,60 +63,19 @@ Razor Pages benefits, in addition to the server rendering benefits:
...
 
-### ASP.NET Core MVC
+## ASP.NET Core MVC
 
-ASP.NET MVC renders UI on the server and uses a Model-View-Controller (MVC) architectural pattern. The MVC pattern separates an app into three main groups of components: Models, Views, and Controllers. User requests are routed to a controller. The controller is responsible for working with the model to perform user actions or retrieve results of queries. The controller chooses the view to display to the user, and provides it with any model data it requires. Support for Razor Pages is built on ASP.NET Core MVC.
+ASP.NET Core MVC is a mature and supported alternative to Blazor. ASP.NET Core MVC renders UI on the server and uses a Model-View-Controller (MVC) architectural pattern. The MVC pattern separates an app into three main groups of components: models, views, and controllers. User requests are routed to a controller. The controller is responsible for working with the model to perform user actions or retrieve results of queries. The controller chooses the view to display to the user and provides it with any model data it requires.
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:68]:
```suggestion
ASP.NET Core MVC renders UI on the server and uses a Model-View-Controller (MVC) architectural pattern. The MVC pattern separates an app into three main groups of components: models, views, and controllers. User requests are routed to a controller. The controller is responsible for working with the model to perform user actions or retrieve results of queries. The controller chooses the view to display to the user and provides it with any model data it requires.
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 14]
```diff
@@ -89,60 +63,19 @@ Razor Pages benefits, in addition to the server rendering benefits:
...
-When a Blazor WebAssembly app is built and run:
-
-* C# code files and Razor files are compiled into .NET assemblies.
-* The assemblies and the [.NET runtime](/dotnet/framework/get-started/overview) are downloaded to the browser.
-* Blazor WebAssembly bootstraps the .NET runtime and configures the runtime to load the assemblies for the app. The Blazor WebAssembly runtime uses JavaScript interop to handle DOM manipulation and browser API calls.
-
-For more information, see <xref:blazor/index> and <xref:blazor/hosting-models>. The server-side Blazor hosting model is described in the [Server-side Blazor](#server-side-blazor) section earlier in this article.
-
-### ASP.NET Core Single Page Application (SPA) with JavaScript Frameworks such as Angular and React
+## ASP.NET Core Single Page Application (SPA) with JavaScript Frameworks such as Angular and React
```
Comment (danroth27) [inline aspnetcore/tutorials/choose-web-ui.md:75]:
```suggestion
## ASP.NET Core Single Page Applications (SPA) with frontend JavaScript frameworks
```

**Review by danroth27 — CHANGES_REQUESTED** [review]
This is a solid start! I have a few comments, particularly on the Blazor section.

File: aspnetcore/tutorials/choose-web-ui.md (hunk 10 above)
Comment (wadepickett) [reply aspnetcore/tutorials/choose-web-ui.md:13]: Commited, thanks.

File: aspnetcore/tutorials/choose-web-ui.md (hunk 11 above)
Comment (wadepickett) [reply aspnetcore/tutorials/choose-web-ui.md]: Removed content on how Blazor can ron on WebAssembly, thanks.

File: aspnetcore/tutorials/choose-web-ui.md (hunk 12 above)
Comment (wadepickett) [reply aspnetcore/tutorials/choose-web-ui.md]: Removed info on how Blazor renders.  Thanks.

File: aspnetcore/tutorials/choose-web-ui.md [hunk 15]
```diff
@@ -3,83 +3,54 @@ title: Choose an ASP.NET Core UI
...
-* Client requirements may exclude users who have low-end devices, older browser versions, or low-bandwidth connections.
+When a Blazor WebAssembly is built and run:
 
-## Choose a server rendered ASP.NET Core UI solution
+* C# code files and Razor files are compiled into .NET assemblies.
+* The assemblies and the [.NET runtime](/dotnet/framework/get-started/overview) are downloaded to the browser.
+* Blazor WebAssembly bootstraps the .NET runtime and configures the runtime to load the assemblies for the app. The Blazor WebAssembly runtime uses JavaScript interop to handle DOM manipulation and browser API calls.
 
-The following section explains the ASP.NET Core web UI server rendered models available and provides links to get started. ASP.NET Core Razor Pages and ASP.NET Core MVC are server-based frameworks for building web apps with .NET.
+For a complete overview Blazor, its architecture and benefits, see <xref:blazor/index> and <xref:blazor/hosting-models>. To get started with your first Blazor app, see [Build your first Blazor app](https://dotnet.microsoft.com/learn/aspnet/blazor-tutorial/intro).
```
Comment (wadepickett) [inline aspnetcore/tutorials/choose-web-ui.md:29]:
```suggestion
For a complete overview of Blazor, its architecture and benefits, see <xref:blazor/index> and <xref:blazor/hosting-models>. To get started with your first Blazor app, see [Build your first Blazor app](https://dotnet.microsoft.com/learn/aspnet/blazor-tutorial/intro).
```

File: aspnetcore/tutorials/choose-web-ui.md [hunk 16]
```diff
@@ -3,83 +3,36 @@ title: Choose an ASP.NET Core UI
...
-* Hybrid apps that take advantage of both server and client UI rendering approaches. For example, most of the web UI is rendered on the server, and client rendered components are added as needed.
+Benefits of using Blazor:
 
-There are benefits and drawbacks to consider when rendering UI on the server or on the client.
+* Reusable component model.
+* Efficient diff-based component rendering.
+* Flexibly render components from the server or client via WebAssembly.
+* Build rich interactive web UI components in C#.
+* Render components statically from the server.
+* Progressively enhance server-rendered components for smoother navigation and form handling and to enable streaming rendering.
```
Comment (wadepickett) [inline aspnetcore/tutorials/choose-web-ui.md:24]:
```suggestion
* Progressively enhance server rendered components for smoother navigation and form handling and to enable streaming rendering.
```

**Comment by wadepickett** [comment]
@danroth27, thanks for the fantastic help.  I think we have all items taken care of now , looking back through each suggestion in the list.  Does it look good to you?

[Internal review link: Choose web UI](https://review.learn.microsoft.com/en-us/aspnet/core/tutorials/choose-web-ui?view=aspnetcore-8.0&branch=pr-en-us-30569)

**Review by danroth27 — APPROVED** [review]
Looks good!

**Pull request merged**